ludwig/datasets/kdd_appetency/__init__.py

```diff
@@ -0,0 +1,134 @@
+        elif filetype == 'txt':
+            file_df = pd.read_csv(
+                os.path.join(self.raw_dataset_path, filename), header=None)
+        return file_df
+
+    def process_downloaded_dataset(self, header=0):
+        super().process_downloaded_dataset()
+        unprocessed_df = pd.read_csv(os.path.join(self.processed_dataset_path,
+                                                self.csv_filename))
+        
+        train_df = unprocessed_df[unprocessed_df[SPLIT] == 0]
+        self.test_df = unprocessed_df[unprocessed_df[SPLIT] == 2]
+        
+        categorical_features = { 190, 191, 192, 193, 194, 195, 196, 197, 198,
+                                199, 200, 201, 202, 203, 204, 205, 206, 207,
+                                209, 210, 211, 212, 213, 214, 215, 216, 217,
+                                218, 219, 220, 221, 222, 223, 224, 225, 226,
+                                227, 228 }
+
+        train_df = process_categorical_features(train_df, categorical_features)
+        train_df = process_numerical_features(train_df, categorical_features)
+        
+        target = self.read_file('labels', os.path.join(self.raw_dataset_path,
+                                                       'orange_small_train_appetency.labels'))
+        
+        train_idx = self.read_file('txt', os.path.join(self.raw_dataset_path,
+                                                       'stratified_train_idx_appetency.txt'))
+        
+        val_idx = self.read_file('txt', os.path.join(self.raw_dataset_path,
+                                                       'stratified_test_idx_appetency.txt'))
+        
+        processed_train_df = train_df.iloc[train_idx[0]].copy()
+        processed_train_df['target'] = target.iloc[train_idx[0]]
+        
+        processed_val_df = train_df.iloc[val_idx[0]].copy()
+        processed_val_df['target'] = target.iloc[val_idx[0]]
+        processed_val_df['split'] = 1
+        
+        self.data_df = pd.concat([processed_train_df, processed_val_df])
+
+    def load_processed_dataset(self, split) -> Tuple[pd.DataFrame,
+                                                     pd.DataFrame,
+                                                     pd.DataFrame]:
+        if split:
+            if SPLIT in self.data_df:
+                training_set = self.data_df[self.data_df[SPLIT] == 0]
+                val_set = self.data_df[self.data_df[SPLIT] == 1]
+            if SPLIT in self.test_df:
+                test_set = self.test_df[self.test_df[SPLIT] == 2]
+            return training_set, test_set, val_set
+        else:
+            raise ValueError("The dataset does not have splits, "
+                             "load with `split=False`")
+
+
+def process_categorical_features(df, categorical_features):
+    for i in categorical_features:
+        df.iloc[:, i].fillna("?", inplace=True) 
+        df.iloc[:, i].apply(lambda x: to_float_str(x))
```

Comment:
why do you need this?

Reply:
I'm not sure about this. Perhaps, I'm missing something but this was part of the preprocessing they followed (c.f. - https://github.com/catboost/benchmarks/blob/master/quality_benchmarks/prepare_appetency_churn_upselling/prepare_appetency_churn_upselling.ipynb).